### ISSUE TITLE
Compliance and Performance Testing

### ISSUE THREAD

**Issue by gardener-robot-ci-1** [gardener/gardener#13]
**Issue by [kubernetes-jenkins](https://github.com/kubernetes-jenkins)**
_Wednesday Jul 19, 2017 at 18:44 GMT_
_Originally opened as https://git.removed/kubernetes-attic/garden-operator/issues/26_

----

# Story
* As provider I want to be sure our shoot clusters are Kubernetes compliant and show a sufficient performance, so that I dont deliver crap to our consumers.
 
# Motivation
We are still in an early phase and need to further validate our seed/shoot approach, before we roll it out.
 
# Acceptance Criteria
- [ ] Manually run, then automate in Jenkins CI:
  - [ ] E2E tests against a shoot cluster pass
  - [ ] Performance and scalability of a shoot cluster is comparable to our kops-based seed clusters or a Tectonic cluster
 
# Resources
- Performance tests: https://github.com/kubernetes/perf-tests
- E2E test infrastructure: https://github.com/kubernetes/test-infra
- E2E testing: https://github.com/kubernetes/community/blob/master/contributors/devel/e2e-tests.md
  According to Martin Vladev <martin.vladev@sap.com> you specify a specific label e.g.:
  ```
  go run hack/e2e.go -- -v --test --test_args="--ginkgo.focus=\[Feature:Performance\]"
  ```
- Maybe [Heptio Ark](https://blog.heptio.com/announcing-two-new-heptio-open-source-projects-heptio-ark-and-heptio-sonobuoy-7cef88a06f8) (disaster recovery) may be interesting to check out?

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?
- [ ] Unit Tests are provided: Have you written automated unit tests or added manual NGPTT tickets?
- [ ] Integration Tests are provided: Have you written automated integration tests?
- [ ] Minimum API exposure: If you have added public API, was it really necessary/is it minimal?
- [ ] Operations guide: Have you updated the [operations guide](https://git.infra.removed/kubernetes/kube-docs/wiki/Operations%20Guide)?

:information_source: Migrated from Jira issue [KUBE-184](https://jtrack.removed/browse/KUBE-184)

**Comment by gardener-robot-ci-1** [comment]
**Comment by [mvladev](https://github.com/mvladev)**
_Wednesday Jul 26, 2017 at 11:32 GMT_

----

The documentation that is available for e2e testing is only valid for local testing scenario (create your own cluster, write code/tests, run & debug them).

What we want is to use the available tests (for a specific k8s release e.g. release-1.6, release-1.7), not modify them and simply run them.

The idea is the following:

For every release (1.5, 1.6, 1.7...) compile `kubectl`, `ginkgo` and `e2e.test` in the kubernetes cross compile image: 
```
#!/bin/bash -eu

# Note: cloning k8s source takes quite some time
# CGO_ENABLED=0 is needed, because e2e.test and ginkgo are not statically compiled by default
# In 1.6 e2e.test can't be statically compiled...
cmd=$(cat <<EOF
  git clone https://github.com/kubernetes/kubernetes.git /go/src/k8s.io/kubernetes --branch v1.6.7 --single-branch \
  && cd /go/src/k8s.io/kubernetes \
  && make all WHAT=cmd/kubectl \
  && CGO_ENABLED=0 make all WHAT=vendor/github.com/onsi/ginkgo/ginkgo \
  && make all WHAT=test/e2e/e2e.test \
  && cp /go/src/k8s.io/kubernetes/_output/bin/{e2e.test,ginkgo,kubectl} /output/
EOF
)
docker run --rm -v $PWD/output:/output -w /output gcr.io/google_containers/kube-cross:v1.7.5-2  /bin/bash -c "${cmd}"

docker build . -t e2e-tests:v1.6
```
Using the following `Dockerfile`
```
# e2e.tests go needs glibc, because it can't statically compile in 1.6.x
# This can be changed to iron/go in 1.7.x
FROM frolvlad/alpine-glibc:alpine-3.6
ADD output/ /usr/local/bin/
WORKDIR /k8s
```
Create a slim image that we will use for e2e tests: `docker build . -t e2e-tests:v1.6.3`

Run conformance tests against`>= 1.6 &&  < 1.7` shoot cluster:
```
#!/bin/bash -eu
# Run e2e conformance tests in parallel
# kubeconfig should be mounted in /config
docker run --rm -v $PWD/config:/config e2e-tests:v1.6  ginkgo \
  --noColor \
  --nodes=25 \
  "/usr/local/bin/e2e.test" -- \
    --kubeconfig=/config/kubeconfig \
    --ginkgo.flakeAttempts=2 \
    --host=https://api.i068969-development.shoot.garden.k8s.sapcloud.io \
    --provider=skeleton \
    --ginkgo.focus="\[Conformance\]" \
    --ginkgo.skip="\[Serial\]"

# [Serial] tests can't be run together with other tests in parallel
docker run --rm -v $PWD/config:/config e2e-tests:v1.6 ginkgo \
    --noColor  \
    "/usr/local/bin/e2e.test" -- \
    --kubeconfig=/config/kubeconfig \
    --ginkgo.flakeAttempts=2 \
    --host=https://api.i068969-development.shoot.garden.k8s.sapcloud.io \
    --provider=skeleton \
    --ginkgo.focus="\[Serial\].*\[Conformance\]"
```

On my machine e2e-conformance testing against a shoot cluster takes ~15 mins

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Wednesday Jul 26, 2017 at 14:58 GMT_

----

That sounds pretty good. Can we integrate this into a (for now) separate Jenkins job (that we will later bring together into/integrate with a CI/CD pipeline)?

One problem will be the Docker part. I have not yet tried to run Docker in Docker, but that's what we will need, right? We have to solve the problem anyhow for the CI/CD pipeline in general, not only for the tests.

About the CI/CD pipeline: we will get support from a colleague from the cloud quality coaching team (CQC). What I hope to achieve is a setup similar to our previous project where we used Jenkins and ran all Jenkins jobs in individual containers. However, the hand-made solution we used back then can't be brought easily to Jenkins on Kubernetes and the Jenkins Kubernetes plugin sucks as it tries to preserve the slave approach, but we want total container isolation on each Jenkins job. Maybe we have to question Jenkins in general, but it is a very general and versatile tool and while there may be better tools for build pipelines (we used Concourse for Cloud Foundry which is closer to containers, but sucks on other levels), Jenkins is the most accepted CI system (but lacks "proper" Kubernetes support).

**Comment by gardener-robot-ci-1** [comment]
**Comment by [mvladev](https://github.com/mvladev)**
_Thursday Jul 27, 2017 at 08:43 GMT_

----

> Note: my previous comment is updated with correct commands

You can expose the docker socket to your Jenkins container and run docker on the node that's running:
```
apiVersion: v1 
kind: Pod 
metadata: 
    name: run-docker 
spec: 
    containers: 
      - name: docker-cmds 
        image: docker:1.12.6 
        command: ['docker', 'run', '--rm', '-p', '80:80', 'httpd:latest'] 
        volumeMounts: 
          - mountPath: /var/run 
            name: docker-sock 
    volumes: 
      - name: docker-sock 
        hostPath: 
            path: /var/run 
```
I'll look at https://github.com/kubernetes/test-infra/tree/master/prow and the whole test infrastructure that k8s runs. I really like the aspect of using github comments on PRs and issues to instruct the CI to run jobs/add labels etc. It might give us some insight.

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Thursday Jul 27, 2017 at 10:55 GMT_

----

Yes, I mounted the socket, libs, and some more stuff into the Jenkins container to implement the isolation for the Service Fabrik CI/CD, but that was easy as I only ran a single Docker node.

Now we have Kubernetes and a.) it doesn't have to be Docker (could be rkt), b.) we would make no use of the cluster as the containers would be spawned on the same node c.) hidden from Kubernetes.

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Saturday Jul 29, 2017 at 05:58 GMT_

----

Side note: Just stumbled across the guideline for the cluster autoscaler on how to run e2e tests before you make a PR to the cluster autoscaler: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#developer. Once we have implemented our procedure, it would be great to trigger it for PRs automaticall (PR builder).

**Comment by gardener-robot-ci-1** [comment]
**Comment by [mvladev](https://github.com/mvladev)**
_Sunday Jul 30, 2017 at 17:47 GMT_

----

On Friday I managed to configure and run [Prow](https://github.com/kubernetes/test-infra/tree/master/prow) on my dev cluster. It's looking very promising. 

The scenario is the following:
* PR is created
* No automatic test are triggered (PR needs to be reviewed first)
* Once a team member reviews it he can trigger e2e-tests with `/test all` comment in the PR
* Job(s) status is added to the PR once the rest succeeds/fails
* Pipeline jobs can be added if needed
* Also I'm running some periodic jobs
* More or less the [official scenario](https://github.com/kubernetes/test-infra/blob/master/prow/architecture.md)

Everything is running in a k8s pods. Prow jobs are TPR and can be created via kubectl

**Comment by gardener-robot-ci-1** [comment]
**Comment by [mvladev](https://github.com/mvladev)**
_Sunday Jul 30, 2017 at 17:51 GMT_

----

Here is a a video for more technical details https://youtu.be/VT7111hp8Zg?t=110

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Sunday Jul 30, 2017 at 19:48 GMT_

----

Thanks, looks interesting, a lightweight solution. The thing I am quite unsure about are our requirements. Overall, building a web app (Gardener UI) and a backend component (Garden Operator) is only one small part of the whole thing. We will need jobs for all kind of things and complicated pipelines, too. Jenkins is quite old and its model doesn't fit Kubernetes at all. On the other hand, there are a ton of plugins out there and its quite flexible.

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Friday Aug 25, 2017 at 04:25 GMT_

----

Added estimate that only refers to the remaining effort. The bulk of the work was already done by Katya and Martin, but we must integrate their work into the CI/CD pipeline.

**Comment by vlerenc** [comment]
Moved to https://github.wdf.sap.corp/kubernetes/landscape-dev-garden/issues/8.
